SignalServiceKit/src/Storage/TSStorageManager.m

```diff
@@ -33,6 +33,91 @@
 static NSString *keychainService          = @"TSKeyChainService";
 static NSString *keychainDBPassAccount    = @"TSDatabasePass";
 
+static BOOL isDatabaseInitializedFlag = NO;
+
+NSObject *isDatabaseInitializedFlagLock()
+{
+    static NSObject *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        instance = [NSObject new];
+    });
+    return instance;
+}
+
+BOOL isDatabaseInitialized()
+{
+    @synchronized(isDatabaseInitializedFlagLock())
+    {
+        return isDatabaseInitializedFlag;
+    }
+}
+
+void setDatabaseInitialized()
+{
+    @synchronized(isDatabaseInitializedFlagLock())
+    {
+        isDatabaseInitializedFlag = YES;
+    }
+}
+
+#pragma mark -
+
+@interface YapDatabaseConnection ()
+
+- (id)initWithDatabase:(YapDatabase *)inDatabase;
+
+@end
+
+#pragma mark -
+
+@interface OWSDatabaseConnection : YapDatabaseConnection
+
+@end
+
+#pragma mark -
+
+@implementation OWSDatabaseConnection
+
+- (void)readWriteWithBlock:(void (^)(YapDatabaseReadWriteTransaction *transaction))block
+{
+    OWSAssert(isDatabaseInitialized());
+
+    [super readWriteWithBlock:block];
+}
+
+@end
+
+#pragma mark -
+
+@interface YapDatabase ()
+
+- (void)addConnection:(YapDatabaseConnection *)connection;
+
+@end
+
+#pragma mark -
+
+@interface OWSDatabase : YapDatabase
+
+@end
+
+#pragma mark -
+
+@implementation OWSDatabase
+
+- (YapDatabaseConnection *)newConnection
```

Comment:
suggested comment:

This clobbers the superclass implementation to include instrumentation which ensures the database is in a ready state before creating transactions.